packages/block-library/src/gallery/gallery-image-styles.native.scss

```diff
@@ -0,0 +1,95 @@
+.container {
+  flex: 1;
+  height: 150px;
```

Comment:
Do we have plans to make this dynamic in another PR?

When I tested the tiles component, I used mostly portrait images and it felt odd that they were cropped on the sides. I think the web implementation is currently better at maintaining aspect ratio when possible, and cropping consistently (our tiles crop things differently depending on position)

![gallery-layout-web-native](https://user-images.githubusercontent.com/8739/68193475-5f9f1200-ffb3-11e9-8830-5709e409aba0.png)

BTW, I think the current layout would be perfectly acceptable for a v1, but I want to make that decision explicit. cc @iamthomasbishop

Reply:
I would agree, the way that the web handles cropping is better in terms of honoring aspect ratios. Obviously, the sooner we can iterate to match this behavior, the better, but I wouldn't consider it a blocker for a first iteration.

Reply:
Thanks for testing this out Koke.
> Do we have plans to make this dynamic in another PR?

I'm glad you asked about that, as I've been contemplating how best to address the aspect ratio concern. For simplicity in this iteration, I've "hard-coded" the height here to a somewhat arbitrary value. I've been wondering what the optimal solution for mobile would be, and in particular, the best way to utilize a common `Image` component that better respects aspect ratio.

It seems that the web implementation relies on the browser to set the height of an image based on its width. One concern I have in mind for mobile is the row height adjustment that happens as a result of the image ordering. On web, when I move a portrait image to a "more crowded" row, its width is reduced, reducing its height, potentially altering the height of the row it was in, or the row it was moved to. Here's an example:

![gallery-web-image-heights](https://user-images.githubusercontent.com/8507675/68271465-504fbf80-00ac-11ea-9128-d88e55b94db0.gif)

When this happens on mobile, I wonder whether this "unpredictability" of row heights would affect the UX, since the position of the mover buttons may change. Another consideration, is that image dimensions may not be known until later, so the row size adjustments can happen in a "laggy" way once the image has been loaded over the network. In any case, I think the cross-platform `Image` component can be very useful for whatever route we take.

Reply:
> I'm glad you asked about that, as I've been contemplating how best to address the aspect ratio concern. For simplicity in this iteration, I've "hard-coded" the height here to a somewhat arbitrary value. I've been wondering what the optimal solution for mobile would be, and in particular, the best way to utilize a common Image component that better respects aspect ratio.

MediaUploadProgress calculates the the height according to the container width actually. But we should be able to set the container width first. This is how it works on Image block as well, as far as I recall... When I check web implementation there's some logic that manually calculates the width of every tile in css.

```
// On mobile and responsive viewports, we allow only 1 or 2 columns at the most.
	& .blocks-gallery-image,
	& .blocks-gallery-item {
		width: calc((100% - #{ $grid-size-large }) / 2);
```

```
	// Beyond mobile viewports, we allow up to 8 columns.
	@include break-small {
		@for $i from 3 through 8 {
			&.columns-#{ $i } .blocks-gallery-image,
			&.columns-#{ $i } .blocks-gallery-item {
				width: calc((100% - #{ $grid-size-large } * #{ $i - 1 }) / #{ $i });
				margin-right: 16px;

				// Rules inside this query are only run by Microsoft Edge.
				// Edge miscalculates `calc`, so we have to add some buffer.
				// See also https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/15637241/
				@supports (-ms-ime-align:auto) {
					width: calc((100% - #{ $grid-size-large } * #{ $i - 1 }) / #{ $i } - 1px);
				}
			}
		}

```

We should put a similar logic in JS side, when screen size is less than a certain amount we should limit the displayed column count by 2, for bigger screens we can let up to 8 as web does. At the end it looks to me like we need to calculate the width of every tile on JS side manually.

Reply:
Hi @pinarol :wave: 
> `MediaUploadProgress` calculates the the height according to the container width

I understand this logic lives in the `calculatePreferedImageSize` function within `ImageSize`, imported as a local dependency of `MediaUploadProgress`. This seems to only be used when `coverUrl` is set, and relies on the `onLayout` prop of its wrapping container View.

> When I check web implementation there's some logic that manually calculates the width of every tile in css.

In the web implementation, `width` is calculated in scss, but the css-calculated `height` property seems to be `auto` for the inner-most relevant container (i.e. the image and some containers may have height set to `100%`, but ultimately, the height is from that `auto` value). My concern around mimicking this on mobile is that the row heights will be varied, and possibly "unpredictable". In the gif above, you can see an example of this behavior. Also, you can see there that the aspect ratio of the pelicans image is not honored very well when it's in the row with more images.

> We should put a similar logic in JS side, when screen size is less than a certain amount we should limit the displayed column count by 2, for bigger screens we can let up to 8 as web does. At the end it looks to me like we need to calculate the width of every tile on JS side manually.

In the current state of this first iteration, this calculation is happening in the Gallery and Tiles components. Shall the individual images be responsible for this behavior instead? Or is this a general comment not specific to this component?

Reply:
> In the web implementation, width is calculated in scss, but the css-calculated height property seems to be auto for the inner-most relevant container (i.e. the image and some containers may have height set to 100%, but ultimately, the height is from that auto value). My concern around mimicking this on mobile is that the row heights will be varied, and possibly "unpredictable". In the gif above, you can see an example of this behavior. Also, you can see there that the aspect ratio of the pelicans image is not honored very well when it's in the row with more images.

I still didn't understand how different rows have different heights on web, because actually there's no such thing as different rows, technically, there's only 1 row and all the tiles are just wrapped to the next line when the container width is not enough. I tried disabling some css attributes like height:auto, but no significant change happened. So it is still a mystery to me, I am guessing it is related with the difference of layout engine on browser. Did you understand exactly what's going on there? 

To do the same thing on mobile I think we'd have to find the max height on a (so-called) row, and would set the height of neighbours accordingly on the same row. But it feels bug prone to me. We can try but I think this deserves a separate iteration on mobile. And I agree, for the first iteration I'd be ok with the current state.

Reply:
> I tried disabling some css attributes like height:auto, but no significant change happened. So it is still a mystery to me, I am guessing it is related with the difference of layout engine on browser. Did you understand exactly what's going on there?

I believe the height on web is determined by the layout engine, and I _think_ the `auto` property is the default, so disabling it from a particular element _might_ result in no change in behavior if the "calculated property" for that element is still `auto` (e.g. from parent, or higher).

Regarding the height of the rows, I believe this is a characteristic of Flexbox which we get for free, and I think this will work the same way for us on mobile (i.e. if various heights are defined on the items within a container, each wrapped row height will be the height of the tallest item in that row).

> We can try but I think this deserves a separate iteration on mobile. And I agree, for the first iteration I'd be ok with the current state.

Sounds good to me :smile:. For the later iteration of that, the cross platform `Image` component that @koke created looks promising as a general solution (with the specific aim of matching image behavior on web), and utilizes the [`aspectRatio`](https://facebook.github.io/react-native/docs/layout-props#aspectratio) prop.

Reply:
> Regarding the height of the rows, I believe this is a characteristic of Flexbox which we get for free, and I think this will work the same way for us on mobile (i.e. if various heights are defined on the items within a container, each wrapped row height will be the height of the tallest item in that row).

I am not understanding how this will going to happen. We should define what we mean by "row", there's only 1 real row in gallery but multiple **perceived** rows. And flex box can't automatically adjust row heights for **perceived** rows. 

> For the later iteration of that, the cross platform Image component that @koke created looks promising as a general solution (with the specific aim of matching image behavior on web), and utilizes the aspectRatio prop.

Technically we already have the aspect ratio, MediaUploadProgress can do the same thing. The cross platform Image component brings better architecture but doesn't solve the problem about setting different heights for different perceived rows.

Reply:
I also think it's flexbox. More specifically [align-items](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox#align-items)

> The initial value for this property is stretch and this is why flex items stretch to the height of the tallest one by default. They are in fact stretching to fill the flex container — the tallest item is defining the height of that.

Reply:
> They are in fact stretching to fill the flex container — the tallest item is defining the height of that.

I understand that they can stretch. But what is the tallest one in our case? The tallest one in gallery? Because all the tiles are in a single container. 

Difference on web is they are stretching to fill the container row by row, as if there are different rows, but actually there aren't. That's the thing I don't get.

Reply:
If there's a simple way of doing this I'd really want to do this in this PR, we don't have to wait for the the cross platform Image component. MediaUploadProgress can tell you the width and height if we pass it the coverUrl property to it.

Reply:
> Difference on web is they are stretching to fill the container row by row, as if there are different rows, but actually there aren't. That's the thing I don't get.

That is exactly what flexbox is doing. From the spec, [`flex-wrap: wrap`](https://drafts.csswg.org/css-flexbox-1/#flex-wrap-property) means the _flex container is multi-line_. And [multi-line containers](https://drafts.csswg.org/css-flexbox-1/#multi-line-flex-container):

> Once content is broken into lines, each line is laid out independently; flexible lengths and the justify-content and align-self properties only consider the items on a single line at a time.
>
> In a multi-line flex container (even one with only a single line), the cross size of each line is the minimum size necessary to contain the flex items on the line (after alignment due to align-self), and the lines are aligned within the flex container with the align-content property. In a single-line flex container, the cross size of the line is the cross size of the flex container, and align-content has no effect. The main size of a line is always the same as the main size of the flex container’s content box

I think this is also possible on Yoga, and Image component might be our only issue there, as this automatic sizing relies on the image being able to calculate its intrinsic size. Flexbox can't adapt to "the tallest one", if the items can't tell their own size.

Reply:
> Once content is broken into lines, each line is laid out independently;.

Aha this explains the web part better, thanks! I hope same thing works on mobile, because I don't see any row containers when I inspect view hierarchy.

> Flexbox can't adapt to "the tallest one", if the items can't tell their own size.

Our images can tell their own sizes if we use MediaUploadProgress with coverUrl, but it will take a bit of time until their real aspectRatio is calculated. So we should put a kind of placeholder with a constant height for each image until we get the real aspect ratio from the url. @mkevins could you try this?

Reply:
Hi @pinarol 
> Our images can tell their own sizes if we use MediaUploadProgress with coverUrl

I've tried using this before and encountered a few issues that seem like blockers (i.e. using the sizes calculated from `ImageSize` via the function as children approach). One issue is that the logic used assumes dimensions are numbers, and not strings / percentages, so it may not be working as expected (e.g. `'100%' < '42%' === true` :see_no_evil: ). Another issue is that the additional views in the hierarchy may be interfering with the layout and `resizeMode` style.

<details>
<summary>Try with coverUrl</summary>

```diff
diff --git a/packages/block-library/src/gallery/gallery-image.native.js b/packages/block-library/src/gallery/gallery-image.native.js
index 666a408..64e0bb2 100644
--- a/packages/block-library/src/gallery/gallery-image.native.js
+++ b/packages/block-library/src/gallery/gallery-image.native.js
@@ -135,7 +135,7 @@ class GalleryImage extends Component {
 		const { compose } = StyleSheet;
 
 		const { isUploadInProgress } = this.state;
-		const { isUploadFailed, retryMessage } = params;
+		const { isUploadFailed, retryMessage, finalWidth, finalHeight } = params;
 		const resizeMode = isCropped ? 'cover' : 'contain';
 
 		const imageStyle = [ styles.image, { resizeMode },
@@ -153,7 +153,7 @@ class GalleryImage extends Component {
 		return (
 			<>
 				<Image
-					style={ imageStyle }
+					style={ [ imageStyle, { width: finalWidth, height: finalHeight } ] }
 					source={ { uri: url } }
 					// alt={ alt }
 					accessibilityLabel={ ariaLabel }
@@ -221,7 +221,7 @@ class GalleryImage extends Component {
 	}
 
 	render() {
-		const { id, isBlockSelected, onRemove } = this.props;
+		const { id, isBlockSelected, onRemove, url } = this.props;
 
 		return (
 			<TouchableWithoutFeedback
@@ -230,7 +230,7 @@ class GalleryImage extends Component {
 			>
 				<View style={ styles.container }>
 					<MediaUploadProgress
-						// coverUrl={ url }
+						coverUrl={ url }
 						mediaId={ id }
 						onUpdateMediaProgress={ this.updateMediaProgress }
 						onFinishMediaUploadWithSuccess={ this.finishMediaUploadWithSuccess }
```

also tried:

```diff
@@ -158,6 +158,8 @@ class GalleryImage extends Component {
 					// alt={ alt }
 					accessibilityLabel={ ariaLabel }
 					ref={ this.bindContainer }
+					width={ finalWidth }
+					height={ finalHeight }
 				/>
 				{ isUploadFailed && (
 					<View style={ styles.uploadFailedContainer }>
```
</details>

<details>
<summary>Screenshot</summary>

<img src="https://user-images.githubusercontent.com/8507675/68556360-b4dc9700-047d-11ea-8063-dd917b210627.png" width="360">
</details>

> Technically we already have the aspect ratio, MediaUploadProgress can do the same thing. The cross platform Image component brings better architecture but doesn't solve the problem about setting different heights for tiles in different perceived rows.

The logic used in `MediaUploadProgress` isn't _quite_ the same thing, since it requires pixel-based width, which we don't have in this case. The row heights are solved nicely by Flexbox, so we shouldn't need anything custom for handling that.

Using Koke's cross-platform Image component seems to be promising for handling the aspect ratio issue: [work in progress patch](https://gist.github.com/mkevins/1ea483269ea3b209c7aee2a0026a94ac), but there are a few things to figure out for that (e.g. alignment within the container, passing `resizeMode`, etc.)

<details>
<summary>Screenshot</summary>

<img src="https://user-images.githubusercontent.com/8507675/68560029-fde71800-048a-11ea-9e6e-08313bef0e95.gif" width="360">
</details>

> We can try but I think this deserves a separate iteration on mobile. And I agree, for the first iteration I'd be ok with the current state.

> If there's a simple way of doing this I'd really want to do this in this PR

I'm not seeing a simple solution, so maybe the best course is to defer to a later PR. WDYT?